solidity/contracts/isms/hook/OPL2ToL1CcipReadIsm.sol

```diff
@@ -0,0 +1,181 @@
+        bytes memory ccipReadCallData = _areWeMessageRecipient(_message)
+            ? abi.encodeWithSignature("getWithdrawalProof(bytes)", _message)
+            : abi.encodeWithSignature(
+                "getFinalizeWithdrawalTx(bytes)",
+                _message
+            );
+
+        revert OffchainLookup(
+            address(this),
+            urls,
+            ccipReadCallData,
+            OPL2ToL1CcipReadIsm.process.selector,
+            _message
+        );
+    }
+
+    /// @dev called by the relayer when the off-chain data is ready
+    function process(
+        bytes calldata _metadata,
+        bytes calldata _message
+    ) external {
+        mailbox.process(_metadata, _message);
+    }
+
+    function verify(
+        bytes calldata _metadata,
+        bytes calldata _message
+    ) external override returns (bool) {
+        if (_areWeMessageRecipient(_message)) {
+            _proveWithdrawal(_metadata, _message);
+        } else {
+            _finalizeWithdrawal(_metadata, _message);
+        }
+
+        return true;
+    }
+
+    /**
+     * @dev no-op handle
+     */
+    function handle(
+        uint32 _origin,
+        bytes32 _sender,
+        bytes calldata _messageBody
+    ) external payable {
+        emit ReceivedMessage(_origin, _sender, msg.value, _messageBody);
+    }
+
+    function interchainSecurityModule()
+        external
+        view
+        override
+        returns (IInterchainSecurityModule)
+    {
+        return IInterchainSecurityModule(address(this));
+    }
+
+    function _areWeMessageRecipient(
+        bytes calldata _message
+    ) internal view returns (bool) {
+        return _message.recipient() == address(this).addressToBytes32();
+    }
+
+    /// @dev We check the withdrawal hash here in order to prevent someone
+    /// DDoS-ing the transfer execution by providing a message relative to a legit withdrawal X
+    /// and a valid proof relative to withdrawal Y
+    function _proveWithdrawal(
+        bytes calldata _metadata,
+        bytes calldata _message
+    ) internal {
+        bytes32 withdrawalHash = abi.decode(_message.body(), (bytes32));
+
+        (
+            IOptimismPortal.WithdrawalTransaction memory _tx,
+            uint256 _disputeGameIndex,
+            IOptimismPortal.OutputRootProof memory _outputRootProof,
+            bytes[] memory _withdrawalProof
+        ) = abi.decode(
+                _metadata,
+                (
+                    IOptimismPortal.WithdrawalTransaction,
+                    uint256,
+                    IOptimismPortal.OutputRootProof,
+                    bytes[]
+                )
+            );
+
+        bytes32 untrustedWithdrawalHash = OPL2ToL1Withdrawal.hashWithdrawal(
+            _tx
+        );
+        if (withdrawalHash != untrustedWithdrawalHash) {
+            revert InvalidWithdrawalHash(
+                untrustedWithdrawalHash,
+                withdrawalHash
+            );
+        }
```

Comment:
I dont think the withdrawal hash is trusted because the `proveWithdrawalTransaction` would revert if the `_tx` was invalid?